### PR TITLE
#24114: Optimize C++ concatenation algorithm

### PR DESCRIPTION
**Pull request by omilyutin-tt**
### Ticket
#24114
#17343

### Problem description
Migration of torch sharding requires optimizing C++ path. Currently there is a big performance gap due to unnecessary copying of data.

In addition, there is an opportunity to parallelize data processing (physical -> logical conversion).

### What's changed
* Re-wrote the algorithm to perform concatenation of multi-dimensional chunks of data in lock step. Similar to how chunking is done.
* Added new utils for xtensor conversion: `XtensorAdapter`, `adapt`, and `AdaptedView`. 
* Re-wrote concatenation implementation to rely on the new multi-dimensional path, switched from `get_device_tensors` function to manipulating `DistributedHostBuffer` directly.
* Added an overload `std::pair<std::vector<T>, Shape> MeshToTensor::compose(const Tensor& tensor) const` - this will be adopted on the `to_torch` conversion path in `pytensor.cpp`.

### Checklist (pending)
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/15935984889)
- [x] [T3K unit tests](https://github.com/tenstorrent/tt-metal/actions/runs/15935987767)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/api/ttnn/distributed/distributed_tensor.hpp [hunk 1]
```diff
@@ -142,7 +142,12 @@ class MeshToTensor {
 
     static MeshToTensor create(const MeshDevice& mesh_device, const MeshComposerConfig& config);
 
-    Tensor compose(const std::vector<Tensor>& tensors) const;
+    // Composes multi-device tensor into a single tensor.
+    Tensor compose(const Tensor& tensor) const;
+
+    // Overload that returns a pair of logical data composed from multi-device tensor and its shape.
```
Comment (bbradelTT) [inline ttnn/api/ttnn/distributed/distributed_tensor.hpp]: "from" -> "of a"

Comment (omilyutin-tt) [reply ttnn/api/ttnn/distributed/distributed_tensor.hpp]: Done, thanks!

File: tt-train/sources/ttml/core/xtensor_utils.hpp [hunk 2]
```diff
@@ -67,7 +67,7 @@ auto xtensor_to_span(const xt::xarray<T>& xtensor) {
 
 template <typename T>
 xt::xarray<T> concat(const std::vector<xt::xarray<T>>& v, size_t axis = 0) {
-    return ttnn::experimental::xtensor::concat(v, axis);
+    return ttnn::experimental::xtensor::concat(v, axis).expr();
```
Comment (omilyutin-tt) [inline tt-train/sources/ttml/core/xtensor_utils.hpp:70]: Unfortunately I had to make a copy here :( 

I don't think we should spend time to optimize this though. The new implementation of sharding / concat supports tt-train use case, provides better performance (no unnecessary copying, parallelized processing), and extends to generic ND case. I am planning to migrate C++ concat to Pytorch and file the bug for tt-train to migrate off the custom implementation. Or do it myself :)

File: ttnn/core/tensor/xtensor/partition.cpp [hunk 3]
```diff
@@ -110,99 +112,128 @@ auto chunk_xexpression(
...
+    const auto& expected_shape = first_expr.shape();
+    const size_t num_dims = first_expr.dimension();
+    for (const auto& expr : expressions) {
+        TT_FATAL(expr.dimension() == num_dims, "All expressions must have the same number of dimensions");
+        TT_FATAL(expr.shape() == first_expr.shape(), "All expressions must have the same shape");
+    }
+    const auto normalized_dims = normalize_dims(dims, num_dims);
+    auto sorted_dims = normalized_dims;
+    std::sort(sorted_dims.begin(), sorted_dims.end());
+    TT_FATAL(std::unique(sorted_dims.begin(), sorted_dims.end()) == sorted_dims.end(), "dims must be unique");
```
[the remark this answers is not in the document]
Comment (omilyutin-tt) [reply ttnn/core/tensor/xtensor/partition.cpp:150]: Equivalent code. Cannot mess with the original order of `normalized_dims`.

File: ttnn/core/distributed/distributed_tensor.cpp [hunk 4]
```diff
@@ -345,47 +366,85 @@ class TensorToMesh::Impl {
...
+            dst_buffer = dst_buffer.transform(
+                [&tensor](const tt::tt_metal::HostBuffer& shard) {
+                    return tt::tt_metal::HostBuffer(Tensor(shard, tensor.tensor_spec()).to_vector<T>());
+                },
+                tt::tt_metal::DistributedHostBuffer::ProcessShardExecutionPolicy::PARALLEL);
+        }
+
+        // Convert shards into a linear buffer of xtensor views.
+        std::vector<experimental::xtensor::AdaptedView<const T>> xtensor_views;
+        xtensor_views.reserve(distribution_shape_.mesh_size());
```
[the remark this answers is not in the document]
Comment (omilyutin-tt) [reply ttnn/core/distributed/distributed_tensor.cpp:406]: No comment here

File: ttnn/api/ttnn/distributed/distributed_tensor.hpp [hunk 5]
```diff
@@ -142,7 +142,12 @@ class MeshToTensor {
 
     static MeshToTensor create(const MeshDevice& mesh_device, const MeshComposerConfig& config);
 
-    Tensor compose(const std::vector<Tensor>& tensors) const;
+    // Composes multi-device tensor into a single tensor.
+    Tensor compose(const Tensor& tensor) const;
+
+    // Overload that returns a pair of logical data composed of a multi-device tensor and its shape.
+    template <typename T>
+    std::pair<std::vector<T>, Shape> compose(const Tensor& tensor) const;
```
Comment (cfjchu) [inline ttnn/api/ttnn/distributed/distributed_tensor.hpp:158]: was this new overload used in this PR, or in the next one w/ pytensor?

File: tt-train/sources/ttml/core/xtensor_utils.hpp (hunk 2 above)
Comment (cfjchu) [reply tt-train/sources/ttml/core/xtensor_utils.hpp:70]: `.expr()` is what adds the extra copy? I thought it's still by ref?

File: ttnn/api/ttnn/distributed/distributed_tensor.hpp [hunk 6]
```diff
@@ -142,7 +142,12 @@ class MeshToTensor {
 
     static MeshToTensor create(const MeshDevice& mesh_device, const MeshComposerConfig& config);
 
-    Tensor compose(const std::vector<Tensor>& tensors) const;
+    // Composes multi-device tensor into a single tensor.
+    Tensor compose(const Tensor& tensor) const;
```
Comment (cfjchu) [inline ttnn/api/ttnn/distributed/distributed_tensor.hpp]: what are the expectations on the input tensor and return tensor? Might be useful to add a comment here

**Review by cfjchu — APPROVED** [review]

File: ttnn/api/ttnn/distributed/distributed_tensor.hpp (hunk 6 above)
Comment (nsextonTT) [reply ttnn/api/ttnn/distributed/distributed_tensor.hpp]: Seconding that request.

**Review by nsextonTT — APPROVED** [review]

File: tt-train/sources/ttml/core/xtensor_utils.hpp (hunk 2 above)
Comment (omilyutin-tt) [reply tt-train/sources/ttml/core/xtensor_utils.hpp:70]: `.expr()` is the reference, but the function returns a `xt::xarray<T>`. So this calls a constructor that copies the data.

File: ttnn/api/ttnn/distributed/distributed_tensor.hpp (hunk 5 above)
Comment (omilyutin-tt) [reply ttnn/api/ttnn/distributed/distributed_tensor.hpp:158]: Needed for pytensor, yeah. Also we need to migrate tt-train. 

For now I use it internally.

File: ttnn/api/ttnn/distributed/distributed_tensor.hpp (hunk 6 above)
Comment (omilyutin-tt) [reply ttnn/api/ttnn/distributed/distributed_tensor.hpp]: Added a comment.

+1 more fundamentally, these APIs that take in a `Tensor` and return a `Tensor` are confusing. Applies to this one and mesh mapper above. I think the creation APIs should really be converting logical data (as span/vector) to `Tensor` distributed across mesh, and vice versa. We should not have a concept of "single device" tensor, and this is something we are working on improving.

File: tt-train/sources/ttml/core/xtensor_utils.hpp (hunk 2 above)
Comment (omilyutin-tt) [reply tt-train/sources/ttml/core/xtensor_utils.hpp:70]: Filed #24339 to migrate tt-train to the new APIs

Comment (ayerofieiev-tt) [reply tt-train/sources/ttml/core/xtensor_utils.hpp:70]: @rfurko-tt @dmakoviichuk-tt fyi

**Review by ayerofieiev-tt — APPROVED** [review]
